kernel/opaqueproof.ml

```diff
@@ -16,19 +16,22 @@ open Mod_subst
 type work_list = (Instance.t * Id.t array) Cmap.t * 
   (Instance.t * Id.t array) Mindmap.t
 
+type cooking_info = {
+  modlist : work_list;
+  abstract : Constr.named_context * Univ.Instance.t * Univ.AUContext.t }
+
 type indirect_accessor = {
   access_proof : DirPath.t -> int -> constr option;
+  access_discharge : cooking_info list -> int -> constr -> constr;
 }
 
-type cooking_info = { 
-  modlist : work_list; 
-  abstract : Constr.named_context * Univ.Instance.t * Univ.AUContext.t }
 type proofterm = (constr * Univ.ContextSet.t) Future.computation
+type universes = int
 type opaque =
   | Indirect of substitution list * DirPath.t * int (* subst, lib, index *)
-  | Direct of cooking_info list * proofterm
+  | Direct of universes * cooking_info list * proofterm
```

Comment:
@ppedrot Can you say how addition of universes here and below relate to the change?

Reply:
You need to know the size of the universe context to be able to abstract over it. In the previous implementation this was hidden via a closure in `discharge_direct_constr`, but we now need to store this information somewhere.